src/main/java/net/neoforged/neoforge/common/crafting/RecipePriorityManager.java

```diff
@@ -0,0 +1,98 @@
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMaps;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.Map;
+import java.util.Optional;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.FileToIdConverter;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.resources.Resource;
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.minecraft.world.item.crafting.Recipe;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class RecipePriorityManager extends SimpleJsonResourceReloadListener<JsonElement> {
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+    public static final Logger LOGGER = LogManager.getLogger();
+
+    private Object2IntMap<ResourceKey<Recipe<?>>> recipePriorities = Object2IntMaps.emptyMap();
+    private static final String folder = "recipe_priorities";
+
+    public RecipePriorityManager() {
+        super(ExtraCodecs.JSON, FileToIdConverter.json(folder));
+    }
+
+    @Override
+    protected Map<ResourceLocation, JsonElement> prepare(ResourceManager resourceManager, ProfilerFiller profilerFiller) {
+        Map<ResourceLocation, JsonElement> map = super.prepare(resourceManager, profilerFiller);
+        for (String namespace : resourceManager.getNamespaces()) {
+            ResourceLocation resourceLocation = ResourceLocation.fromNamespaceAndPath(namespace, "recipe_priorities/recipe_priorities.json");
+            Optional<Resource> resource = resourceManager.getResource(resourceLocation);
+            if (resource.isPresent()) {
+                try (Reader reader = resource.get().openAsReader()) {
+                    JsonObject jsonobject = GsonHelper.fromJson(GSON, reader, JsonObject.class);
+                    boolean replace = GsonHelper.getAsBoolean(jsonobject, "replace", false);
+                    if (replace)
+                        map.clear();
+                    JsonObject entries = GsonHelper.getAsJsonObject(jsonobject, "entries");
+                    for (String entry : entries.keySet()) {
+                        ResourceLocation loc = ResourceLocation.tryParse(entry);
+                        map.remove(loc); //remove and re-add if needed, to update the ordering.
```

Comment:
You need a linked hash map for that to work. But I don't see why it would matter.